### PR TITLE
VM tests with OCS disks

### PR DESCRIPTION
**Pull request by ahadas**
**What this PR does / why we need it**:
This PR introduces VM start/stop/delete and serial console connection with OCS disks, if Ceph-RBD is present within the cluster.

**Release note**:

```release-note
Enhance the functional tests with running basic VM operations using OCS disks.
```

### CODE REVIEW

File: tests/utils.go [hunk 1]
```diff
@@ -3666,6 +3669,19 @@ func HasCDI() bool {
 	return HasFeature("DataVolumes")
 }
 
+func HasCeph() (bool, string) {
```
Comment (vladikr) [inline tests/utils.go]: I'm just wondering. Is this only a ceph block storage?
Or can ceph be filesystem as well? If it can, how do we distinguish here?

Comment (ahadas) [reply tests/utils.go]: Specifically on kubevirtci, seems that at the moment [only ceph block storage may be deployed](https://github.com/kubevirt/kubevirtci/blob/master/cluster-provision/k8s/scripts/ceph-csi.sh#L16). That's what I see also in a cluster in which `enable-ceph` is set

**Comment by ahadas** [comment]
@slintes it looks good - the OCS tests are executed within `pull-kubevirt-e2e-k8s-1.15.1-ceph` and skipped in other lanes

File: tests/vm_test.go [hunk 2]
```diff
@@ -92,13 +93,13 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
-			newVMI := NewRandomVirtualMachine(template, false)
-			newVMI.TypeMeta = v12.TypeMeta{
+			newVM := NewRandomVirtualMachine(template, false)
+			newVM.TypeMeta = v12.TypeMeta{
 				APIVersion: v1.GroupVersion.String(),
 				Kind:       "VirtualMachine",
 			}
 
-			result := virtClient.RestClient().Post().Resource("virtualmachines").Namespace(tests.NamespaceTestDefault).Body(newVMI).Do()
+			result := virtClient.RestClient().Post().Resource("virtualmachines").Namespace(tests.NamespaceTestDefault).Body(newVM).Do()
```
Comment (slintes) [inline tests/vm_test.go]: again not really related... but there is a nicer function in virtClient for creating a VM :)

File: tests/vm_test.go [hunk 3]
```diff
@@ -116,86 +117,104 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 
-			var newVMI *v1.VirtualMachine
+			var newVM *v1.VirtualMachine
 			var err error
 
-			newVMI = NewRandomVirtualMachineWithRunStrategy(template, runStrategy)
+			newVM = NewRandomVirtualMachineWithRunStrategy(template, runStrategy)
 
-			newVMI, err = virtClient.VirtualMachine(tests.NamespaceTestDefault).Create(newVMI)
+			newVM, err = virtClient.VirtualMachine(tests.NamespaceTestDefault).Create(newVM)
```
Comment (slintes) [inline tests/vm_test.go]: related to my last comment: like this :)

File: tests/vm_test.go [hunk 4]
```diff
@@ -64,13 +65,13 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 		It("[test_id:1518]should be rejected on POST", func() {
 			vmiImage := tests.ContainerDiskFor(tests.ContainerDiskCirros)
 			template := tests.NewRandomVMIWithEphemeralDiskAndUserdata(vmiImage, "echo Hi\n")
-			newVMI := NewRandomVirtualMachine(template, false)
-			newVMI.TypeMeta = v12.TypeMeta{
+			newVM := NewRandomVirtualMachine(template, false)
+			newVM.TypeMeta = v12.TypeMeta{
```
[the remark this answers is not in the document]
Comment (ahadas) [reply tests/vm_test.go]: Done, removed NewRandomVirtualMachine altogether and use tests.NewRandomVirtualMachine that includes this instead

File: tests/vm_test.go (hunk 2 above)
Comment (ahadas) [reply tests/vm_test.go]: Right, and it's used elsewhere in this module but note that this particular test checks the response of the POST call so I think it's ok here

**Comment by ahadas** [comment]
/hold

missing:
* lifecycle+console tests with volumeMod = File
* remove data volume on test-cleanup
* wait for data volumes to be imported in lifecycle+console test

File: tests/vm_test.go (hunk 2 above)
Comment (slintes) [reply tests/vm_test.go]: oh, I missed that! Then it's ok of course

File: tests/utils.go (hunk 1 above)
Comment (ahadas) [reply tests/utils.go]: Changed `GetCephStorageClass` and added tests for both Ceph filesystem and block volumes in vm/console tests

File: pkg/virt-handler/vm.go [hunk 5]
```diff
@@ -1249,18 +1249,13 @@ func (d *VirtualMachineController) checkVolumesForMigration(vmi *v1.VirtualMachi
 			} else if err != nil {
 				return blockMigrate, err
 			}
-			blockMigrate = blockMigrate || !shared
 			if !shared {
-				return blockMigrate, fmt.Errorf("cannot migrate VMI with non-shared PVCs")
+				return true, fmt.Errorf("cannot migrate VMI with non-shared PVCs")
```
Comment (slintes) [inline pkg/virt-handler/vm.go]: Are we fixing a bug here? Makes sense to me, but would like to have @vladikr have a look at this.

File: tests/console_test.go [hunk 6]
```diff
@@ -75,6 +78,13 @@ var _ = Describe("[rfe_id:127][posneg:negative][crit:medium][vendor:cnv-qe@redha
 		return expecter, errChan
 	}
 
+	deleteDataVolume := func(dv *cdiv1.DataVolume) {
+		if dv != nil {
+			By("g the DataVolume")
```
Comment (slintes) [inline tests/console_test.go]: what does "g" stand for?

File: tests/migration_test.go [hunk 7]
```diff
@@ -1053,17 +1052,54 @@ var _ = Describe("[rfe_id:393][crit:high[vendor:cnv-qe@redhat.com][level:system]
...
+				sc, exists := tests.GetCephStorageClass()
+				if !exists {
+					Skip("Skip OCS tests when Ceph is not present")
+				}
+
+				By("Starting an iSCSI POD")
+				iscsiIP := tests.CreateISCSITargetPOD(tests.ContainerDiskFedora)
+				volMode := k8sv1.PersistentVolumeBlock
+				// create a new PV and PVC (PVs can't be reused)
+				pvName := "test-nfs" + rand.String(48)
```
Comment (slintes) [inline tests/migration_test.go]: nit: this is not nfs

File: tests/console_test.go (hunk 6 above)
Comment (ahadas) [reply tests/console_test.go]: Oops, should be "Deleting" - I'll fix that

File: pkg/virt-handler/vm.go (hunk 5 above)
Comment (ahadas) [reply pkg/virt-handler/vm.go]: No, it's not a bug fix, the following part just confused me:
```golang
blockMigrate = blockMigrate || !shared
if !shared {
  return blockMigrate, fmt.Errorf("cannot migrate VMI with non-shared PVCs")
}
```
so I tried to simplify it:
if `!shared == true` then blockMigrate becomes `true` and we get to the return statement
if `!shared == false` then blockMigrate doesn't change and so the line `blockMigrate = blockMigrate || !shared` can be removed
and then we end up with just:
```golang
if !shared {
  return true, fmt.Errorf("cannot migrate VMI with non-shared PVCs")
}
```

File: tests/migration_test.go (hunk 7 above)
Comment (ahadas) [reply tests/migration_test.go]: Done

File: tests/console_test.go (hunk 6 above)
Comment (ahadas) [reply tests/console_test.go]: Done

File: pkg/virt-handler/vm.go (hunk 5 above)
Comment (slintes) [reply pkg/virt-handler/vm.go]: ah yes, sure... :+1:

**Comment by slintes** [comment]
I didn't see this failure before, appeared on both os-3.11.0 and os-3.11.0-crio.
`VM Live Migration Starting a VirtualMachineInstance with a shared ISCSI Filesystem PVC should migrate a vmi with UNSAFE_MIGRATION flag set`

Also k8s 1.11 has some more failures than usual. Let's retry...

/retest

**Comment by ahadas** [comment]
> New changes are detected. LGTM label has been removed.

Just resolved conflicts in vm_test.go :)

**Comment by ahadas** [comment]
extracted the migration stuff yesterday as the failures seemed related to migration tests but apparently they still fail but specifically on k8s 1.11.0 which I didn't check locally, checking now

**Comment by ahadas** [comment]
Seems like the issue in the tests was fixed - I accidentally removed PV creation that is needed in one of the tests...

**Pull request merged**